core/src/main/java/org/elasticsearch/action/support/ActiveShardsObserver.java

```diff
@@ -0,0 +1,136 @@
+                        // configured number of active shards to be allocated before returning, as that is when indexing
+                        // operations can take place on the newly created index
+                        if (waitForActiveShards == ActiveShardCount.NONE) {
+                            // not waiting, so just run whatever we were to run when the waiting is
+                            actionListener.onResponse(onResult.apply(true, false));
+                            return;
+                        }
+
+                        // wait for the configured number of active shards to be allocated before returning
+                        final ClusterStateObserver observer =
+                            new ClusterStateObserver(clusterService, logger, threadPool.getThreadContext());
+                        final ClusterStateObserver.ChangePredicate shardsAllocatedPredicate = new ClusterStateObserver.ChangePredicate() {
+                            @Override
+                            public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                                                 ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+                                return waitForActiveShards.enoughShardsActive(newState, indexName);
+                            }
+
+                            @Override
+                            public boolean apply(ClusterChangedEvent changedEvent) {
+                                return waitForActiveShards.enoughShardsActive(changedEvent.state(), indexName);
+                            }
+                        };
+
+                        final ClusterStateObserver.Listener observerListener = new ClusterStateObserver.Listener() {
+                            @Override
+                            public void onNewClusterState(ClusterState state) {
+                                actionListener.onResponse(onResult.apply(true, false));
+                            }
+
+                            @Override
+                            public void onClusterServiceClose() {
+                                logger.debug("[{}] cluster service closed while waiting for enough shards to be started.", indexName);
+                                actionListener.onFailure(new NodeClosedException(clusterService.localNode()));
+                            }
+
+                            @Override
+                            public void onTimeout(TimeValue timeout) {
+                                actionListener.onResponse(onResult.apply(true, true));
+                            }
+                        };
+
+                        observer.waitForNextChange(observerListener, shardsAllocatedPredicate, timeout);
+
+                    } catch (Exception ex) {
+                        logger.debug("[{}] index creation failed on waiting for shards", request.index());
```

Comment:
Upon investigation, the only exception that can be thrown is the ElasticsearchException if the cluster state is already being observed by the observer.  This observer is only used within the confines of this method, and is only called here once, so it should not be possible for this exception to be thrown.  However, I feel this try catch block is relevant anyway because if an exception happens to be thrown by the call to waitForNextChange (or some logic is added to the ClusterStateObserver that enables an exception to be thrown), then at least we wouldn't block on the failure here and would send it immediately to the client.